hedera-node/hedera-app/src/main/java/com/hedera/node/app/config/ConfigProviderImpl.java

```diff
@@ -105,142 +106,62 @@ public class ConfigProviderImpl implements ConfigProvider {
+    public static final String APPLICATION_PROPERTIES_DEFAULT_PATH = "application.properties";
+    /** Default path to the semantic-version.properties file. */
+    private static final String SEMANTIC_VERSION_PROPERTIES_DEFAULT_PATH = "semantic-version.properties";
     /**
-     * Default path to the application.properties file.
+     * The actual underlying versioned configuration provided by this provider. This must provided thread-safe access
+     * since many threads (ingestion, pre-handle, etc.) will be accessing it from different threads while it is
+     * generally only updated on the handle thread (except during startup). The handle thread will also access it.
      */
-    public static final String APPLICATION_PROPERTIES_DEFAULT_PATH = "application.properties";
-
-    private volatile VersionedConfiguration configuration;
-
+    private final AtomicReference<VersionedConfiguration> configuration;
+    /** Provides synchronous access to updating the configuration to one thread at a time. */
     private final AutoClosableLock updateLock = Locks.createAutoLock();
 
     /**
-     * Constructor.
+     * Create a new instance. You must specify whether to use the genesis.properties file as a source for the
+     * configuration. This should only be true if the node is starting from genesis.
      */
     @Inject
-    public ConfigProviderImpl(@NonNull @GenesisUsage final Boolean useGenesisSource) {
-        Objects.requireNonNull(useGenesisSource, "useGenesisSource must not be null");
-        ConfigurationBuilder builder = ConfigurationBuilder.create();
-        addConfigData(builder);
-        addConverter(builder);
-        addValidators(builder);
+    public ConfigProviderImpl(@GenesisUsage final Boolean useGenesisSource) {
+        final var builder = newConfigurationBuilder();
         addFileSources(builder, useGenesisSource);
         final Configuration config = builder.build();
-        configuration = new VersionedConfigImpl(config, 0);
+        configuration = new AtomicReference<>(new VersionedConfigImpl(config, 0));
+    }
+
+    @Override
+    @NonNull
+    public VersionedConfiguration getConfiguration() {
+        return configuration.get();
     }
 
     /**
      * This method must be called if a property has changed. It will update the configuration and increase the version.
-     * This should happen whenever {@link GlobalDynamicProperties#reload()} is called.
      *
      * @param propertyFileContent the new property file content
      */
     public void update(@NonNull final Bytes propertyFileContent) {
-        try (final var lock = updateLock.lock()) {
-            ConfigurationBuilder builder = ConfigurationBuilder.create();
-            addConfigData(builder);
-            addConverter(builder);
-            addValidators(builder);
+        try (final var ignoredLock = updateLock.lock()) {
+            final var builder = newConfigurationBuilder();
+            addFileSources(builder, false);
```

Comment:
Maybe I got it wrong but I understood that the `Bytes` will contain all properties and therefore the files should be ignored once we receive an update.

Reply:
Right, I think that was incorrect. There are 3 different types of properties:

1. Network properties. These are properties that affect the network as a whole. Every node in the network must have the same properties or they will potentially ISS. These properties start off in `genesis.properties` but only for the genesis use case, otherwise they are all in File 121 (during genesis the file service loads the genesis properties into the file 121).
2. Node properties. These are properties that affect the node only, and can be totally different on different nodes. For example, the directory in which to store database files, etc. These are never stored in File 121, because File 121 has to be consistent for all nodes, and these properties, by definition, may not be consistent on all nodes.
3. semantic-version.properties. These properties contain the version numbers of the deployed software.

When File 121 is reloaded, we still need the properties system to include `semantic-version.properties` and the node properties, in addition to File 121. Only `genesis.properties` would be ignored.

We do not yet, but should, support ENV and system properties as additional "node property" sources, but they should not ever have the ability to override the File 121 properties (if they did, an ISS is almost inevitable, so we'd like to foreclose on that possibility so we don't accidentally mess it up).